server/services/streaming_sequence_service.go

```diff
@@ -0,0 +1,178 @@
+			codes.InvalidArgument,
+			"The field `name` is required.")
+	}
+
+	// Retrieve Sequence and associated SequenceReport.
+	i, ok := s.streamingsequences.Load(name)
+	if !ok {
+		return status.Errorf(
+			codes.NotFound,
+			"The Sequence with %q does not exist.",
+			name,
+		)
+	}
+	seq := i.(*pb.StreamingSequence)
+
+	i, _ = s.streamingreports.Load(streaming_report(name))
+	rep, _ := i.(*pb.StreamingSequenceReport)
+
+	// Get the number of attempts, which coincides with this attempt's number.
+	n := len(rep.Attempts)
+
+	// Prepare the attempt response defined by the Sequence.
+	st := status.New(codes.OK, "Successful attempt")
+	sendStatusAtIndex := 0
+	var delay time.Duration
+	responses := seq.GetResponses()
+	content := strings.Fields(seq.GetContent())
+	if l := len(responses); l > 0 && n < l {
+		resp := responses[n]
+		delay = resp.GetDelay().AsDuration()
+		st = status.FromProto(resp.GetStatus())
+		sendStatusAtIndex = int(resp.SendStatusAtIndex) - len(strings.Fields(s.sent_content))
+
+		if s.sent_content != "" {
+			words_written := strings.Fields(s.sent_content)
+			content = content[len(words_written):]
+		}
+
+	} else if n > l {
+		st = status.New(codes.OutOfRange, "Attempt exceeded predefined responses")
+	}
+
+	for idx, word := range content {
+		if idx >= sendStatusAtIndex {
+			break
+		}
+		s.sent_content += word + " "
+		err := stream.Send(&pb.AttemptStreamingSequenceResponse{Content: word})
+		if err != nil {
+			return err
+		}
+	}
+
+	echoStreamingHeaders(stream)
+
+	echoStreamingTrailers(stream)
+
+	time.Sleep(delay)
+
+	// Calculate the perceived delay since the last RPC attempt.
+	attDelay := &duration.Duration{}
+	if n > 0 {
+		prev := rep.GetAttempts()[n-1]
+		respTime := prev.GetResponseTime()
+		d := received.Sub(respTime.AsTime())
+		attDelay = ptypes.DurationProto(d)
+	}
+
+	// Clock the time that the server is sending the response
+	responseTime := time.Now()
+	rpb, err := ptypes.TimestampProto(responseTime)
+	if err != nil {
+		return status.Errorf(
+			codes.Internal,
+			err.Error(),
+		)
+	}
+
+	rep.Attempts = append(rep.Attempts, &pb.StreamingSequenceReport_Attempt{
+		AttemptNumber: int32(n),
+		ResponseTime:  rpb,
+		AttemptDelay:  attDelay,
+		Status:        st.Proto(),
+		ContentSent:   s.sent_content,
+	})
+
+	if n+1 >= len(responses) {
+		s.sent_content = ""
+	}
+
+	return st.Err()
+
+}
+
+func (s *sequenceServerImpl) GetStreamingSequenceReport(ctx context.Context, in *pb.GetStreamingSequenceReportRequest) (*pb.StreamingSequenceReport, error) {
+	name := in.GetName()
+	if name == "" {
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"The field `name` is required.")
+	}
+
+	report, ok := s.streamingreports.Load(name)
+	if !ok {
+		return nil, status.Errorf(
+			codes.NotFound,
+			"The sequence report with %q does not exist.",
+			name,
+		)
+	}
+
+	return report.(*pb.StreamingSequenceReport), nil
+}
+
+func clone_streamingSequence(s *pb.StreamingSequence) *pb.StreamingSequence {
```

Comment:
I think you can avoid all this with a single [proto.Clone](https://pkg.go.dev/google.golang.org/protobuf/proto#Clone)?

Reply:
Just for more context, I think @galz10 was looking to replicate the logic for the [unary version] of this method. We can def try replace with proto.Clone in one or both spots if appropriate - up to @noahdietz here

Reply:
tbh I don't remember why I wrote a bespoke `clone` method but in looking at the implementation I think it is because `proto.Clone` wasn't doing a recursive clone for other proto messages used in the message e.g. the `repeated Response responses` field. The nature of the in-memory database that Showcase uses means we have to be very particular in how we copy b.c the client libraries writing test code against the same spots in memory might modify the data and screw something up.

Reply:
Given this and that it seems to be working, I'm inclined to leave it as is for now!

Reply:
(leaving the comment unresolved for @galz10 to read when he's back, Gal, resolve when you've read it!)